src/jumps.jl

```diff
@@ -50,28 +50,57 @@ jump process is
+vrj = VariableRateJump(rate, affect!; lrate=lrate, urate=urate, L=L)
 ```
 
 ## Notes
-- **`VariableRateJump`s result in `integrator`s storing an effective state type
-  that wraps the main state vector.** See [`ExtendedJumpArray`](@ref) for
-  details on using this object. Note that the presence of *any*
-  `VariableRateJump`s will result in all `ConstantRateJump`, `VariableRateJump`
-  and callback `affect!` functions receiving an integrator with `integrator.u`
-  an [`ExtendedJumpArray`](@ref).
-- Must be used with `ODEProblem`s or `SDEProblem`s to be correctly simulated
-  (i.e. can not currently be used with `DiscreteProblem`s).
+- When using the `Coevolve` aggregator, `DiscreteProblem` can be used.
+  Otherwise, `ODEProblem` or `SDEProblem` must be used to be correctly simulated.
+- **When not using the `Coevolve` aggregator, `VariableRateJump`s result in
+  `integrator`s storing an effective state type that wraps the main state
+  vector.** See [`ExtendedJumpArray`](@ref) for details on using this object. Note
+  that the presence of *any* `VariableRateJump`s will result in all
+  `ConstantRateJump`, `VariableRateJump` and callback `affect!` functions
+  receiving an integrator with `integrator.u` an [`ExtendedJumpArray`](@ref).
 - Salis H., Kaznessis Y.,  Accurate hybrid stochastic simulation of a system of
   coupled chemical or biochemical reactions, Journal of Chemical Physics, 122
   (5), DOI:10.1063/1.1835951 is used for calculating jump times with
   `VariableRateJump`s within ODE/SDE integrators.
 """
-struct VariableRateJump{R, F, I, T, T2} <: AbstractJump
-    """Function `rate(u,p,t)` that returns the jump's current rate."""
+struct VariableRateJump{R, F, R2, R3, R4, I, T, T2} <: AbstractJump
+    """Function `rate(u,p,t)` that returns the jump's current rate given state
+    `u`, parameters `p` and time `t`."""
     rate::R
-    """Function `affect(integrator)` that updates the state for one occurrence of the jump."""
+    """Function `affect!(integrator)` that updates the state for one occurrence
+    of the jump given `integrator`."""
     affect!::F
+    """When planning to use the `Coevolve` aggregator, function `lrate(u, p,
+    t)` that computes the lower bound of the rate in interval `t` to `t + L` at time
+    `t` given state `u`, parameters `p`. This is not required if using another
+    aggregator."""
+    lrate::R2
+    """When planning to use the `Coevolve` aggregator, function `urate(u, p,
+    t)` that computes the upper bound of the rate in interval `t` to `t + L` at time
+    `t` given state `u`, parameters `p`. This is not required if using another
+    aggregator."""
+    urate::R3
+    """When planning to use the `Coevolve` aggregator, function `L(u, p,
+    t)` that computes the interval  length `L` starting at time `t` given state
+    `u`, parameters `p` for which the rate is bounded between `lrate` and
+    `urate`. This is not required if using another aggregator."""
+    L::R4
```

Comment:
Also, we should think a bit about interface. If `L` is going to require some subset of `urate`, `lrate` and/or the actual rate, we might want to have a way to (optionally) pass them to `L` if the algorithm has valid values for them (avoiding the need to call them extra times). For example, we could require `L` to support passing them in as optional keywords (with default values of `nothing`). In this way the algorithm can just pass in those that it knows, and leave any others the user may need for their custom `L` function to calculate.

Alternatively, would it make sense to merge all three functions into one that returns the upper/lower bounds and the time window?

Reply:
I did some experimentation with different APIs including the one you suggested of having a single function that returns the upper/lower bounds and the time window. I eventually settled for this one, as I thought it could be simpler and more modular. Since defining `L` is more art, than science. The user could play with different definitions of `L` without touching `urate` or `lrate`. Also, there might be aggregators that do need  either one of `L`, `urate` or `lrate`.

Previously, I also had problems with defining `FunctionWrapper`s for complex outputs. But then, I also tried incorporating closures and event history which was the main reason why I struggled with `FunctionWrapper`s. Maybe this time I will get luckier.

`Coevolve` will evaluate `urate` and `L` right at the beginning of the main loop. While `lrate`, is optional and only evaluated later. Ideally, to avoid re-evaluating functions, the user could defined memoized functions. The [`Memoization.jl`](https://github.com/marius311/Memoization.jl) library could be an option for this. Internally, we might want to consider wrapping these functions with the `@memoize` macro, but I'm not sure whether this would bring the same efficiency gains as using the keyword argument.

Alternatively, I would propose that we require two functions `urate` and `lrate`. `urate` returns the upper-bound and the time-interval; while `lrate` is optional and takes `urate, and `L` as optional keywords to avoid re-evaluation. The main drawback is the lack of modularity with the potential gain in efficiency if not using memoization.

Let me know what you think.

Reply:
I think keeping them separate is fine for now, we can add the optional kwargs with cached values later own that might help reduce performance overhead.

Reply:
Ok, then I will mark this conversation as resolved.